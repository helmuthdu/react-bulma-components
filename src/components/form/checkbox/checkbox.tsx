import cn from 'classnames';
import React from 'react';
import { Colors, Sizes } from '../../../constants';
import modifiers, { ElementModifier } from '../../../modifiers';

type CheckboxProps = ElementModifier & {
  color?: Colors;
  indeterminate?: boolean;
  size?: Sizes;
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'color' | 'size' | 'unselectable' | 'value'>;

export const Checkbox: React.FunctionComponent<CheckboxProps> = ({
  checked,
  children,
  className,
  color,
  disabled,
  indeterminate,
  name,
  size,
  style,
  value,
  testId,
  ...rest
}: CheckboxProps) => {
  const props = modifiers.clean(rest);
  return (
    <label
      data-testid={(Array.isArray(testId) ? testId[0] : testId) || undefined}
      className={cn('b-checkbox checkbox', modifiers.getClassName(rest), className)}
      style={style}
    >
      <input
        data-testid={Array.isArray(testId) ? testId[2] : undefined}
        checked={checked}
        disabled={disabled}
        indeterminate={indeterminate}
        name={name}
        type="checkbox"
        value={value}
        {...props}
      />
      <span
        data-testid={Array.isArray(testId) ? testId[1] : undefined}
        className={cn('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Checkbox.defaultProps = {
  ...modifiers.defaultProps,
  checked: false,
  indeterminate: false
};
