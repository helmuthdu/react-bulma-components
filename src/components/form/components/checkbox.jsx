import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Checkbox = ({
  checked,
  children,
  className,
  color,
  disabled,
  indeterminate,
  name,
  size,
  style,
  value,
  ...allProps
}) => {
  const props = modifiers.clean(allProps);
  return (
    <label
      disabled={disabled}
      className={cn('b-checkbox checkbox', modifiers.getClassName(allProps), className)}
      style={style}
      data-testid="checkbox-label"
    >
      <input
        {...props}
        name={name}
        type="checkbox"
        value={value}
        disabled={disabled}
        checked={checked}
        indeterminate={indeterminate}
        data-testid="checkbox-input"
      />
      <span
        className={cn('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
        data-testid="checkbox-check"
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Checkbox.propTypes = {
  ...modifiers.propTypes,
  checked: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  indeterminate: PropTypes.bool,
  name: PropTypes.string,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  value: PropTypes.string
};

Checkbox.defaultProps = {
  ...modifiers.defaultProps,
  checked: false,
  indeterminate: false,
  value: ''
};
