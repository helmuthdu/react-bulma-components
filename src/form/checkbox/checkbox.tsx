import clsx from 'clsx';
import * as React from 'react';
import { Colors, Sizes } from '../../constants';
import modifiers, { ElementModifier } from '../../modifiers';

type CheckboxProps = ElementModifier & {
  color?: Colors;
  indeterminate?: boolean;
  size?: Sizes;
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'ref' | 'color' | 'size' | 'unselectable' | 'value'>;

export const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  ({ checked, children, className, color, disabled, indeterminate, name, size, style, value, ...props }, ref) => (
    <label className={clsx('checkbox-container checkbox', modifiers.getClassName(props), className)} style={style}>
      <input
        ref={ref}
        checked={checked}
        disabled={disabled}
        indeterminate={indeterminate}
        name={name}
        type="checkbox"
        value={value}
        {...modifiers.clean(props)}
      />
      <span
        className={clsx('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  )
);

Checkbox.defaultProps = {
  ...modifiers.defaultProps,
  checked: false,
  indeterminate: false
};
