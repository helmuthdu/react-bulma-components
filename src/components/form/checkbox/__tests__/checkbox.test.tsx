import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Checkbox } from '../checkbox';

describe('Checkbox component', () => {
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render', () => {
    const { asFragment } = render(<Checkbox onChange={onChange}>Text</Checkbox>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render unchecked', () => {
    const { getByTestId } = render(<Checkbox onChange={onChange} />);
    expect(getByTestId('checkbox-input')).not.toHaveAttribute('checked', '');
  });

  it('should set input checked if checked', () => {
    const { getByTestId } = render(<Checkbox onChange={onChange} checked />);
    expect(getByTestId('checkbox-input')).toHaveAttribute('checked', '');
  });

  it('should change value on change event', () => {
    const { getByTestId } = render(<Checkbox onChange={onChange}>Text</Checkbox>);
    fireEvent.click(getByTestId('checkbox-label'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
