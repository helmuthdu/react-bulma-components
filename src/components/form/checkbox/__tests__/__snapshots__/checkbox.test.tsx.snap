// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render 1`] = `
<DocumentFragment>
  <label
    class="b-checkbox checkbox"
  >
    <input
      type="checkbox"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;
